File: examples/go/projection_example.go
// Field Projection Examples for Go Client
//
// Demonstrates how to use SelectFields and ExcludeFields to optimize
// data transfer and improve query performance.
//
// Shows comparison between full object queries and projected queries.

package main

import (
	"fmt"
	"log"
	"os"
	"sort"
	"time"

	ekodb "github.com/ekoDB/ekodb-client-go"
	"github.com/joho/godotenv"
)

const testCollection = "projection_test_users_go"

func main() {
	// Load environment variables
	if err := godotenv.Load(); err != nil {
		log.Println("No .env file found")
	}

	baseURL := os.Getenv("API_BASE_URL")
	if baseURL == "" {
		baseURL = "http://localhost:8080"
	}

	apiKey := os.Getenv("API_BASE_KEY")
	if apiKey == "" {
		apiKey = "a-test-api-key-from-ekodb"
	}

	// Initialize client
	client, err := ekodb.NewClient(baseURL, apiKey)
	if err != nil {
		log.Fatalf("Failed to create client: %v", err)
	}

	fmt.Println("Client created\n")

	// Setup: Create test collection and insert test data
	fmt.Println("Setting up test data...")
	// Cleanup any existing test collection
	_ = client.DeleteCollection(testCollection)

	// Insert test users with various fields
	testUsers := []ekodb.Record{
		{
			"name":         "Alice Johnson",
			"email":        "alice@example.com",
			"age":          30,
			"status":       "active",
			"user_role":    "admin",
			"password":     "secret123",
			"api_key":      "key123",
			"secret_token": "token123",
			"bio":          "Software engineer",
			"avatar_url":   "https://example.com/alice.jpg",
			"created_at":   time.Now().Format(time.RFC3339),
		},
		{
			"name":         "Bob Smith",
			"email":        "bob@example.com",
			"age":          25,
			"status":       "active",
			"user_role":    "user",
			"password":     "secret456",
			"api_key":      "key456",
			"secret_token": "token456",
			"bio":          "Designer",
			"avatar_url":   "https://example.com/bob.jpg",
			"created_at":   time.Now().Format(time.RFC3339),
		},
		{
			"name":         "Carol White",
			"email":        "carol@example.com",
			"age":          35,
			"status":       "inactive",
			"user_role":    "user",
			"password":     "secret789",
			"api_key":      "key789",
			"secret_token": "token789",
			"bio":          "Manager",
			"avatar_url":   "https://example.com/carol.jpg",
			"created_at":   time.Now().Format(time.RFC3339),
		},
		{
			"name":         "Dave Brown",
			"email":        "dave@example.com",
			"age":          45,
			"status":       "active",
			"user_role":    "admin",
			"password":     "secretabc",
			"api_key":      "keyabc",
			"secret_token": "tokenabc",
			"bio":          "CTO",
			"avatar_url":   "https://example.com/dave.jpg",
			"created_at":   time.Now().Format(time.RFC3339),
		},
	}

	var insertedIDs []string
	for _, user := range testUsers {
		result, err := client.Insert(testCollection, user)
		if err != nil {
			log.Printf("Warning: Insert failed: %v", err)
		} else if id, ok := result["id"].(string); ok {
			insertedIDs = append(insertedIDs, id)
		}
	}
	fmt.Printf("Inserted %d test users\n\n", len(insertedIDs))

	// Example 1: Select specific fields - only get id, name, email
	fmt.Println("Example 1: Select specific fields (id, name, email only)")
	usersQuery := ekodb.NewQueryBuilder().
		Eq("status", "active").
		SelectFields("id", "name", "email").
		Limit(10).
		Build()

	users, err := client.Find(testCollection, usersQuery)
	if err != nil {
		log.Fatalf("Find failed: %v", err)
	}
	fmt.Printf("  Found %d active users\n", len(users))
	if len(users) > 0 {
		fields := getFieldNames(users[0])
		fmt.Printf("  Fields returned: %v\n", fields)
		name := ekodb.GetStringValue(users[0]["name"])
		email := ekodb.GetStringValue(users[0]["email"])
		fmt.Printf("  First user: %s <%s>\n", name, email)
	}

	// Example 2: Exclude sensitive fields - hide password, api_key, secret_token
	fmt.Println("\nExample 2: Exclude sensitive fields (password, api_key, secret_token)")
	adminsQuery := ekodb.NewQueryBuilder().
		Eq("user_role", "admin").
		ExcludeFields("password", "api_key", "secret_token").
		Build()

	admins, err := client.Find(testCollection, adminsQuery)
	if err != nil {
		log.Fatalf("Find failed: %v", err)
	}
	fmt.Printf("  Found %d admins\n", len(admins))
	if len(admins) > 0 {
		_, hasPassword := admins[0]["password"]
		_, hasApiKey := admins[0]["api_key"]
		_, hasToken := admins[0]["secret_token"]
		fmt.Println("  Sensitive fields excluded:")
		fmt.Printf("    - password: %s\n", excludedStatus(hasPassword))
		fmt.Printf("    - api_key: %s\n", excludedStatus(hasApiKey))
		fmt.Printf("    - secret_token: %s\n", excludedStatus(hasToken))
		fields := getFieldNames(admins[0])
		fmt.Printf("  Fields returned: %v\n", fields)
	}

	// Example 3: Complex query with projection - active users with profile fields
	fmt.Println("\nExample 3: Complex query with projection (active users, ages 18-65)")
	activeUsersQuery := ekodb.NewQueryBuilder().
		Eq("status", "active").
		Gte("age", 18).
		Lt("age", 65).
		SelectFields("id", "name", "email", "age", "created_at").
		SortDescending("age").
		Limit(50).
		Build()

	activeUsers, err := client.Find(testCollection, activeUsersQuery)
	if err != nil {
		log.Fatalf("Find failed: %v", err)
	}
	fmt.Printf("  Found %d active users (ages 18-65)\n", len(activeUsers))
	for _, user := range activeUsers {
		name := ekodb.GetStringValue(user["name"])
		age, _ := ekodb.GetIntValue(user["age"])
		fmt.Printf("    - %s (age %d)\n", name, age)
	}

	// Example 4: Query inactive users with profile fields
	fmt.Println("\nExample 4: Query inactive users with profile fields")
	inactiveQuery := ekodb.NewQueryBuilder().
		Eq("status", "inactive").
		SelectFields("id", "name", "email", "bio").
		Build()

	inactiveUsers, err := client.Find(testCollection, inactiveQuery)
	if err != nil {
		log.Fatalf("Find failed: %v", err)
	}
	fmt.Printf("  Found %d inactive users\n", len(inactiveUsers))
	for _, user := range inactiveUsers {
		name := ekodb.GetStringValue(user["name"])
		bio := ekodb.GetStringValue(user["bio"])
		fmt.Printf("    - %s: %s\n", name, bio)
	}

	// Example 5: Compare full vs projected data - demonstrates bandwidth savings
	fmt.Println("\nExample 5: Compare full vs projected data")
	fullQuery := ekodb.NewQueryBuilder().
		Eq("status", "active").
		Build()

	projectedQuery := ekodb.NewQueryBuilder().
		Eq("status", "active").
		SelectFields("id", "name", "email").
		Build()

	fullUsers, err := client.Find(testCollection, fullQuery)
	if err != nil {
		log.Fatalf("Find failed: %v", err)
	}

	projectedUsers, err := client.Find(testCollection, projectedQuery)
	if err != nil {
		log.Fatalf("Find failed: %v", err)
	}

	if len(fullUsers) > 0 && len(projectedUsers) > 0 {
		fullFields := getFieldNames(fullUsers[0])
		projectedFields := getFieldNames(projectedUsers[0])

		fmt.Println("  Full query:")
		fmt.Printf("    - %d fields per record\n", len(fullFields))
		fmt.Printf("    - Fields: %v\n", fullFields)
		fmt.Println("  Projected query:")
		fmt.Printf("    - %d fields per record\n", len(projectedFields))
		fmt.Printf("    - Fields: %v\n", projectedFields)
		savings := 100 - (len(projectedFields) * 100 / max(len(fullFields), 1))
		fmt.Printf("  Bandwidth savings: ~%d%% fewer fields\n", savings)
	}

	// Cleanup
	fmt.Println("\nCleaning up test data...")
	if err := client.DeleteCollection(testCollection); err != nil {
		log.Printf("Warning: Failed to delete collection: %v", err)
	} else {
		fmt.Println("Cleanup complete")
	}

	fmt.Println("\nAll projection examples completed successfully!")
}

// Helper function to get sorted field names from a record
func getFieldNames(record ekodb.Record) []string {
	fields := make([]string, 0, len(record))
	for k := range record {
		fields = append(fields, k)
	}
	sort.Strings(fields)
	return fields
}

// Helper function to return exclusion status string
func excludedStatus(present bool) string {
	if present {
		return "PRESENT (unexpected!)"
	}
	return "excluded"
}

// Helper function for max
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: examples/rust/examples/projection_example.rs
//! Field Projection Examples for Rust Client
//!
//! Demonstrates how to use select_fields and exclude_fields to optimize
//! data transfer and improve query performance.
//!
//! Shows comparison between full object queries and projected queries.

use ekodb_client::{Client, FieldType, QueryBuilder, Record, Result};
use std::env;

const TEST_COLLECTION: &str = "projection_test_users_rs";

#[tokio::main]
async fn main() -> Result<()> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    let client = Client::builder()
        .base_url(&base_url)
        .api_key(&api_key)
        .build()?;

    println!("Client created\n");

    // Setup: Create test collection and insert test data
    println!("Setting up test data...");
    // Try to delete existing collection (ignore error if doesn't exist)
    let _ = client.delete_collection(TEST_COLLECTION).await;

    // Insert test users with various fields
    let test_users = vec![
        Record::new()
            .field("name", "Alice Johnson")
            .field("email", "alice@example.com")
            .field("status", "active")
            .field("user_role", "admin")
            .field("age", 30i64)
            .field("bio", "Software engineer")
            .field("avatar_url", "https://example.com/alice.jpg")
            .field("password", "secret123")
            .field("api_key", "key123")
            .field("secret_token", "token123")
            .field("created_at", chrono::Utc::now().to_rfc3339()),
        Record::new()
            .field("name", "Bob Smith")
            .field("email", "bob@example.com")
            .field("status", "active")
            .field("user_role", "user")
            .field("age", 25i64)
            .field("bio", "Designer")
            .field("avatar_url", "https://example.com/bob.jpg")
            .field("password", "secret456")
            .field("api_key", "key456")
            .field("secret_token", "token456")
            .field("created_at", chrono::Utc::now().to_rfc3339()),
        Record::new()
            .field("name", "Carol White")
            .field("email", "carol@example.com")
            .field("status", "inactive")
            .field("user_role", "user")
            .field("age", 35i64)
            .field("bio", "Manager")
            .field("avatar_url", "https://example.com/carol.jpg")
            .field("password", "secret789")
            .field("api_key", "key789")
            .field("secret_token", "token789")
            .field("created_at", chrono::Utc::now().to_rfc3339()),
        Record::new()
            .field("name", "Dave Brown")
            .field("email", "dave@example.com")
            .field("status", "active")
            .field("user_role", "admin")
            .field("age", 45i64)
            .field("bio", "CTO")
            .field("avatar_url", "https://example.com/dave.jpg")
            .field("password", "secretabc")
            .field("api_key", "keyabc")
            .field("secret_token", "tokenabc")
            .field("created_at", chrono::Utc::now().to_rfc3339()),
    ];

    let mut inserted_ids: Vec<String> = Vec::new();
    for user in test_users {
        let result = client.insert(TEST_COLLECTION, user, None).await?;
        if let Some(FieldType::String(id)) = result.get("id") {
            inserted_ids.push(id.clone());
        }
    }
    println!("Inserted {} test users\n", inserted_ids.len());

    // Example 1: Select specific fields - only get id, name, email
    println!("Example 1: Select specific fields (id, name, email only)");
    let users_query = QueryBuilder::new()
        .eq("status", "active")
        .select_fields(vec![
            "id".to_string(),
            "name".to_string(),
            "email".to_string(),
        ])
        .limit(10)
        .build();

    let users = client.find(TEST_COLLECTION, users_query, None).await?;
    println!("  Found {} active users", users.len());
    if !users.is_empty() {
        let fields: Vec<&String> = users[0].keys().collect();
        println!("  Fields returned: {:?}", fields);
        // Show first user's data
        if let Some(FieldType::String(name)) = users[0].get("name") {
            if let Some(FieldType::String(email)) = users[0].get("email") {
                println!("  First user: {} <{}>", name, email);
            }
        }
    }

    // Example 2: Exclude sensitive fields - hide password, api_key, secret_token
    println!("\nExample 2: Exclude sensitive fields (password, api_key, secret_token)");
    let admins_query = QueryBuilder::new()
        .eq("user_role", "admin")
        .exclude_fields(vec![
            "password".to_string(),
            "api_key".to_string(),
            "secret_token".to_string(),
        ])
        .build();

    let admins = client.find(TEST_COLLECTION, admins_query, None).await?;
    println!("  Found {} admins", admins.len());
    if !admins.is_empty() {
        let has_password = admins[0].get("password").is_some();
        let has_api_key = admins[0].get("api_key").is_some();
        let has_token = admins[0].get("secret_token").is_some();
        println!("  Sensitive fields excluded:");
        println!(
            "    - password: {}",
            if has_password {
                "PRESENT (unexpected!)"
            } else {
                "excluded"
            }
        );
        println!(
            "    - api_key: {}",
            if has_api_key {
                "PRESENT (unexpected!)"
            } else {
                "excluded"
            }
        );
        println!(
            "    - secret_token: {}",
            if has_token {
                "PRESENT (unexpected!)"
            } else {
                "excluded"
            }
        );
        let fields: Vec<&String> = admins[0].keys().collect();
        println!("  Fields returned: {:?}", fields);
    }

    // Example 3: Complex query with projection - active users with profile fields
    println!("\nExample 3: Complex query with projection (active users, ages 18-65)");
    let active_users_query = QueryBuilder::new()
        .eq("status", "active")
        .gte("age", 18)
        .lt("age", 65)
        .select_fields(vec![
            "id".to_string(),
            "name".to_string(),
            "email".to_string(),
            "age".to_string(),
            "created_at".to_string(),
        ])
        .sort_desc("age")
        .limit(50)
        .build();

    let active_users = client
        .find(TEST_COLLECTION, active_users_query, None)
        .await?;
    println!("  Found {} active users (ages 18-65)", active_users.len());
    for user in &active_users {
        if let (Some(FieldType::String(name)), Some(FieldType::Integer(age))) =
            (user.get("name"), user.get("age"))
        {
            println!("    - {} (age {})", name, age);
        }
    }

    // Example 4: Query with projection by status
    println!("\nExample 4: Query inactive users with profile fields");
    let inactive_query = QueryBuilder::new()
        .eq("status", "inactive")
        .select_fields(vec![
            "id".to_string(),
            "name".to_string(),
            "email".to_string(),
            "bio".to_string(),
        ])
        .build();

    let inactive_users = client.find(TEST_COLLECTION, inactive_query, None).await?;
    println!("  Found {} inactive users", inactive_users.len());
    for user in &inactive_users {
        if let (Some(FieldType::String(name)), Some(FieldType::String(bio))) =
            (user.get("name"), user.get("bio"))
        {
            println!("    - {}: {}", name, bio);
        }
    }

    // Example 5: Compare full vs projected data - demonstrates bandwidth savings
    println!("\nExample 5: Compare full vs projected data");
    let full_query = QueryBuilder::new().eq("status", "active").build();

    let projected_query = QueryBuilder::new()
        .eq("status", "active")
        .select_fields(vec![
            "id".to_string(),
            "name".to_string(),
            "email".to_string(),
        ])
        .build();

    let full_users = client.find(TEST_COLLECTION, full_query, None).await?;
    let projected_users = client.find(TEST_COLLECTION, projected_query, None).await?;

    if !full_users.is_empty() && !projected_users.is_empty() {
        let full_fields: Vec<&String> = full_users[0].keys().collect();
        let projected_fields: Vec<&String> = projected_users[0].keys().collect();

        println!("  Full query:");
        println!("    - {} fields per record", full_fields.len());
        println!("    - Fields: {:?}", full_fields);
        println!("  Projected query:");
        println!("    - {} fields per record", projected_fields.len());
        println!("    - Fields: {:?}", projected_fields);
        println!(
            "  Bandwidth savings: ~{}% fewer fields",
            100 - (projected_fields.len() * 100 / full_fields.len().max(1))
        );
    }

    // Cleanup
    println!("\nCleaning up test data...");
    client.delete_collection(TEST_COLLECTION).await?;
    println!("Cleanup complete");

    println!("\nAll projection examples completed successfully!");

    Ok(())
}
